src/plugins/unified_field_list/public/hooks/use_grouped_fields.ts

```diff
@@ -0,0 +1,270 @@
+
+  const unfilteredFieldGroups: FieldListGroups<T> = useMemo(() => {
+    const containsData = (field: T) => {
+      if (!dataViewId || !dataView) {
+        return true;
+      }
+      const overallField = dataView.getFieldByName?.(field.name);
+      return Boolean(overallField && hasFieldDataHandler(dataViewId, overallField.name));
+    };
+
+    const fields = allFields || [];
+    const allSupportedTypesFields = onSupportedFieldFilter
+      ? fields.filter(onSupportedFieldFilter)
+      : fields;
+    const sortedFields = [...allSupportedTypesFields].sort(sortFields);
+    const groupedFields = {
+      ...getDefaultFieldGroups(),
+      ...groupBy(sortedFields, (field) => {
+        if (field.type === 'document') {
+          return 'specialFields';
+        } else if (dataView?.metaFields?.includes(field.name)) {
+          return 'metaFields';
+        } else if (containsData(field)) {
+          return 'availableFields';
+        } else return 'emptyFields';
+      }),
+    };
+    const selectedFields = onSelectedFieldFilter ? sortedFields.filter(onSelectedFieldFilter) : [];
+
+    let fieldGroupDefinitions: FieldListGroups<T> = {
+      SpecialFields: {
+        fields: groupedFields.specialFields,
+        fieldCount: groupedFields.specialFields.length,
+        isAffectedByGlobalFilter: false,
+        isAffectedByTimeFilter: false,
+        isInitiallyOpen: false,
+        showInAccordion: false,
+        title: '',
+        hideDetails: true,
+      },
+      SelectedFields: {
+        fields: selectedFields,
+        fieldCount: selectedFields.length,
+        isInitiallyOpen: true,
+        showInAccordion: true,
+        title: i18n.translate('unifiedFieldList.useGroupedFields.selectedFieldsLabel', {
+          defaultMessage: 'Selected fields',
+        }),
+        isAffectedByGlobalFilter: false,
+        isAffectedByTimeFilter: true,
+        hideDetails: false,
+        hideIfEmpty: true,
+      },
+      AvailableFields: {
+        fields: groupedFields.availableFields,
+        fieldCount: groupedFields.availableFields.length,
+        isInitiallyOpen: true,
+        showInAccordion: true,
+        title:
+          dataViewId && fieldsExistenceInfoUnavailable
+            ? i18n.translate('unifiedFieldList.useGroupedFields.allFieldsLabel', {
+                defaultMessage: 'All fields',
+              })
+            : i18n.translate('unifiedFieldList.useGroupedFields.availableFieldsLabel', {
+                defaultMessage: 'Available fields',
+              }),
+        helpText: i18n.translate('unifiedFieldList.useGroupedFields.allFieldsLabelHelp', {
+          defaultMessage: 'Data view fields', // TODO: what text should be in here by default?
+        }),
+        isAffectedByGlobalFilter: false,
+        isAffectedByTimeFilter: true,
+        // Show details on timeout but not failure
+        // hideDetails: fieldsExistenceInfoUnavailable && !existenceFetchTimeout, // TODO: is this check still necessary?
```

Comment:
The question is how to catch timeouts of [`loadFieldExisting`](https://github.com/elastic/kibana/blob/0b1218cdf428de7208e9af5b6a54361f4aa95d38/src/plugins/unified_field_list/public/services/field_existing/load_field_existing.ts) after it was moved from Lens server API to public of unifiedFieldList plugin in https://github.com/elastic/kibana/pull/139453. It might be not relevant anymore.

cc @dimaanj

Reply:
I also think that this is no longer necessary since dataViewsService `getFieldsForIndexPattern` doesn't support a dedicated timeout error state, right @mattkime ?